include/swift/IDE/CodeCompletion.h

```diff
@@ -612,30 +618,420 @@ enum class CompletionKind {
+/// it to either of those.
+/// Context-free diagnostics are independent of the context they are used in.
+/// The not-recommended reason can thus be cached as part of a context free
+/// code completion result.
+/// Contextual not recommended reasons depend on the context they are used in.
+/// E.g. \c InvalidAsyncContext depends on whether the usage context is async or
+/// not.
+enum class NotRecommendedReason : uint8_t {
+  None = 0,                    // both contextual and context-free
+  RedundantImport,             // contextual
+  RedundantImportIndirect,     // contextual
+  Deprecated,                  // context-free
+  SoftDeprecated,              // context-free
+  InvalidAsyncContext,         // contextual
+  CrossActorReference,         // contextual
+  VariableUsedInOwnDefinition, // contextual
+
+  MAX_VALUE = VariableUsedInOwnDefinition
+};
+
+/// TODO: We consider deprecation warnings as context free although they don't
+/// produce deprecation warnings insdide context that have the same or a
+/// stronger deprecation attribute.
+/// E.g.
+/// In SDK:
+/// \code
+/// @available(iOS, deprecated: 12.0)
+/// func deprecatedFunc()
+/// \endcode
+///
+/// User code:
+/// \code
+/// @available(iOS, deprecated: 12.0)
+/// func insiderUserDeprecatedContext() {
+///   #^COMPLETE^#
+/// }
+/// \endcode
+/// suggests \c deprecatedFunc as deprecated although it doesn't produce a
+/// diagnostic during compilation. But this allows us to show deprecation
+/// warnings for cached results.
+enum class ContextFreeNotRecommendedReason : uint8_t {
+  None = 0,
+  Deprecated,
+  SoftDeprecated,
+  MAX_VALUE = SoftDeprecated
+};
+
+enum class ContextualNotRecommendedReason : uint8_t {
+  None = 0,
+  RedundantImport,
+  RedundantImportIndirect,
+  InvalidAsyncContext,
+  CrossActorReference,
+  VariableUsedInOwnDefinition,
+  MAX_VALUE = VariableUsedInOwnDefinition
+};
+
+enum class CodeCompletionResultKind : uint8_t {
+  Declaration,
+  Keyword,
+  Pattern,
+  Literal,
+  BuiltinOperator,
+
+  MAX_VALUE = BuiltinOperator
+};
+
+/// The type returned by a code completion item. It is canonicalized and can be
+/// instantiated in different \c ASTContexts.
+class CodeCompletionResultType {
+
+  /// The USR of the type. Used to re-instantiate the type in different \c
+  /// ASTContexts.
+  std::string USR;
+
+  /// If a type has already been computed from the USR, it is cached here.
+  Type Ty;
+
+  /// The \c ASTContext in which \c Ty lives. Must be stored separately from
+  /// \c Ty because the \c ASTContext in which \c Ty lives might get destroyed,
+  /// thus making the pointer inside \c Ty invalid.
+  ASTContext *ContextOfTy;
+
+  /// Maps USRs to \c CodeCompletionResultTypes that have already been
+  /// constructed.
+  static llvm::StringMap<CodeCompletionResultType *> USRCache;
+
+  /// Maps Types to \c CodeCompletionResultTypes that have already been
+  /// constructed.
+  static llvm::DenseMap<Type, CodeCompletionResultType *> TypeCache;
```

Comment:
Oh, I thought that cache would only be introduced by the next PR, sorry. 

I will store them in a dedicate arena with https://github.com/ahoppen/swift/commit/e9ef698fa91625f027221a403e58b691d2bfe9a8 but because that change is >200 LOC as well, I didn’t want to inflate this PR even further. Do you think we should pull that commit into this PR or should we just fix it in a follow-up PR?

We can’t store `CodeCompletionResultType` in the `ASTContext` because they are used for cached results, which are used for multiple code completion sessions and thus outlive the `ASTContext`.

Reply:
> We can’t store CodeCompletionResultType in the ASTContext because they are used for cached results, which are used for multiple code completion sessions and thus outlive the ASTContext.

I'm not sure the cache should outlive the ASTContext. An ASTContext is used for multiple code completion sessions as long as fast-completion is working. If fast-completion is not working and the ASTContext is destroyed, I think the cache should be cleared, or what cache can we reuse?